src/compiler/scala/tools/nsc/ast/parser/Scanners.scala

```diff
@@ -705,14 +728,18 @@ trait Scanners extends ScannersCommon {
                   syntaxError("empty character literal (use '\\'' for single quote)")
                 else {
                   nextChar()
-                  token = CHARLIT
-                  setStrVal()
+                  if (Character.isHighSurrogate(cbuf.charAt(0)))
```

Comment:
To keep accepting `val high = '\ud801'` (and also the more questionable literal version of that sequence), maybe something like this?

```suggestion
                  if (cbuf.length == 2)
```

Reply:
Indeed, it's important to allow `'\uhhhh'` escapes for surrogate chars. Also to allow unpaired surrogate chars in `\uhhhh` form within string literals. Both are important to write tests of string-related operations that need to deal with this stuff.

Reply:
Invalid unicode string literals are unaffected by this PR and are still accepted, it's just the unpaired high surrogate escape char literals.

If this goes through `\Uhhhhhh` escapes in addition to `\uhhhh` escapes would also be neat.

Reply:
Thanks, that crossed my mind but I wasn't sure of the use case or test, see the new test. That relaxes both the `\u` example and having an orphan surrogate in input. Note that under the old unicode regime, escape was processed early and the two cases were not distinguishable. Also noticed backquoted idents are literal chars that fall under the purview.